repository/src/main/java/org/apache/atlas/discovery/EntityLineageService.java

```diff
@@ -466,7 +469,9 @@ private boolean incrementAndCheckIfRelationsLimitReached(AtlasEdge atlasEdge, bo
             outLineageInfo.setHasMoreOutputs(true);
             hasRelationsLimitReached = true;
         } else {
-            outLineageInfo.incrementOutputRelationsCount();
+            if (! (direction.equals(AtlasLineageOnDemandInfo.LineageDirection.INPUT) && outGuid.equals(baseGuid))) {
```

Comment:
Is it possible to have a self-cyclic relation at direction OUTPUT here?

Reply:
so, when there is a self-cyclic lineage, it is affecting both INPUT and OUTPUT side. but, since we are processing INPUT at first when it is asked for direction BOTH, we need to not increase the count for the output side since it is self-cyclic. When we are processing for OUTPUT side, it does not matter because the issue here the fact that INPUT side affecting OUTPUT limits.

Reply:
From the description:

> When the first query's direction is set as BOTH and the second as OUTPUT, remaining vertices in the second page does not get retrieved.
> 
> Encountered issue:
> 
> An entity has many self-cyclic processes and many output vertices linked with a single process. At first, the graph is loaded with direction as BOTH. And, when show more clicked on the output side, the direction is set as OUTPUT and from becomes 10 for the next page. What happens is, self-cyclic relations are loaded at first when the direction is BOTH and when we switch to OUTPUT, we specify the OUTPUT direction's next page. But we omit what was there in the first page of OUTPUT.